subsys/net/lib/latmon/latmon.c

```diff
@@ -0,0 +1,458 @@
+			continue;
+		}
+
+		if (send_sample_data(msg->latmus, data) < 0) {
+			break;
+		}
+	}
+
+	k_sem_give(&monitor_done);
+	monitor_tid = NULL;
+
+	LOG_INF("\tmonitoring stopped");
+
+	return 0;
+}
+
+static int get_latmus_conf(ssize_t len, struct latmon_net_request *req,
+			struct latmus_conf *conf)
+{
+	if (len != sizeof(*req)) {
+		return -1;
+	}
+
+	if (!ntohl(req->period_usecs)) {
+		LOG_ERR("null period received, invalid\n");
+		return -1;
+	}
+
+	if (ntohl(req->period_usecs) > MAX_SAMPLING_PERIOD_USEC) {
+		LOG_ERR("invalid period received: %u usecs\n",
+			ntohl(req->period_usecs));
+		return -1;
+	}
+
+	if (ntohl(req->histogram_cells) > HISTOGRAM_CELLS_MAX) {
+		LOG_ERR("invalid histogram size received: %u > %u cells\n",
+			ntohl(req->histogram_cells), HISTOGRAM_CELLS_MAX);
+		return -1;
+	}
+
+	conf->period = ntohl(req->period_usecs);
+	conf->cells = ntohl(req->histogram_cells);
+	conf->max_samples = MAX_SAMPLING_PERIOD_USEC / conf->period;
+
+	return 0;
+}
+
+static void start_monitoring(struct latmon_message *msg,
+			struct latmus_conf *conf,
+			struct latmon_data *data)
+{
+	k_sem_reset(&monitor_done);
+	abort_monitor = false;
+
+	memset(data, 0, sizeof(*data));
+	monitor_tid = k_thread_create(&monitor_thread, monitor_stack,
+			MONITOR_STACK_SIZE, (k_thread_entry_t)monitor_entry,
+			msg, conf, data,
+			MONITOR_THREAD_PRIORITY, 0, K_NO_WAIT);
+}
+
+static void stop_monitoring(void)
+{
+	if (!monitor_tid) {
+		return;
+	}
+
+	abort_monitor = true;
+	k_sem_take(&monitor_done, K_FOREVER);
+}
+
+static void handle_connection(struct latmon_message *msg)
+{
+#if (K_HEAP_MEM_POOL_SIZE > 0)
+	struct latmus_conf *conf = k_malloc(sizeof(*conf));
+	struct latmon_data *data = k_malloc(sizeof(*data));
+	struct latmon_net_request req;
+	ssize_t len;
+
+	if (!conf || !data) {
+		LOG_ERR("Failed to allocate memory, check HEAP_MEM_POOL_SIZE");
+		goto out;
+	}
+
+	memset(conf, 0, sizeof(*conf));
+	memset(data, 0, sizeof(*data));
+	for (;;) {
+		len = recv(msg->latmus, &req, sizeof(req), 0);
+		stop_monitoring();
+		if (get_latmus_conf(len, &req, conf) < 0) {
+			/* we might have to send the histogram */
+			if (send_trailing_data(msg->latmus, conf, data) < 0) {
+				break;
+			}
+			memset(conf, 0, sizeof(*conf));
+			continue;
+		}
+		start_monitoring(msg, conf, data);
+	}
+out:
+	k_free(conf);
+	k_free(data);
+	close(msg->latmus);
+	k_sem_give(&latmon_done);
+#else
+	LOG_ERR("No heap configured");
+#endif
+}
+
+static int latmon_server_thread(void *p1, void *p2, void *p3)
+{
+	struct latmon_message msg = { };
+
+	for (;;) {
+		k_msgq_get(&latmon_msgq, &msg, K_FOREVER);
+
+		/* Only latmus can stop the monitoring, so hang in there */
+		handle_connection(&msg);
+	}
+
+	return 0;
+}
+
+static void broadcast_ip_address(struct in_addr *ip_addr)
+{
+	char ip_str[NET_IPV4_ADDR_LEN];
+	struct sockaddr_in broadcast;
+	int sock;
+
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		perror("socket");
+		exit(1);
```

Comment:
Please avoid POSIX functions in kernel subsystems. Not entirely sure if this would even compile on a non posix target even with POSIX_API selected, or what it would do :) Just return -1 and propagate it to the caller of `net_latmon_connect()`